fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/MixinRegistry.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.registry.sync;
+
+import java.util.EnumSet;
+import java.util.Set;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.event.registry.RegistryAttribute;
+import net.fabricmc.fabric.api.event.registry.RegistryAttributeHolder;
+import net.fabricmc.fabric.impl.registry.sync.FabricRegistry;
+import net.fabricmc.fabric.impl.registry.sync.HashedRegistry;
+
+@Mixin(Registry.class)
+public abstract class MixinRegistry<T> implements RegistryAttributeHolder, FabricRegistry, HashedRegistry {
+	@Unique
+	private final EnumSet<RegistryAttribute> attributes = EnumSet.noneOf(RegistryAttribute.class);
+
+	@Unique
+	private int preBootstrapHash = -1;
+
+	@Shadow
+	public abstract Set<Identifier> getIds();
+
+	/**
+	 * This is used to denote backwards compatibility, when false default attributes will be applied.
+	 * This is set to true when using FabricRegistryBuilder
+	 */
+	@Unique
+	private boolean builtWithBuilder = false;
+
+	@Override
+	public RegistryAttributeHolder addAttribute(RegistryAttribute attribute) {
+		attributes.add(attribute);
+		return this;
+	}
+
+	@Override
+	public boolean hasAttribute(RegistryAttribute attribute) {
+		return attributes.contains(attribute);
+	}
+
+	@Override
+	public void build(Set<RegistryAttribute> attributes) {
+		builtWithBuilder = true;
+		this.attributes.addAll(attributes);
+	}
+
+	@Override
+	public boolean builtByBuilder() {
+		return builtWithBuilder;
+	}
+
+	@Override
+	public int getStoredHash() {
+		return preBootstrapHash;
+	}
+
+	@Override
+	public int storeHash() {
+		return preBootstrapHash = getIds().hashCode();
```

Comment:
This implementation of `storeHash` rings alarm bells for me. This `hashCode` is implemented by `AbstractSet`, which simply sums the hashCode of its elements. Each element is an `Identifier`, which is itself a bad hash function.
Rather than using a hash, it would be far better to simply check the size of the registry to check if it has been added to. No-one is going to remove from a registry after all (that's a hard thing to do and no-one would want to anyway).

Reply:
Removal from registries to replace stuff has been seen before, the hallow did it to allow place able pumpkin pie.

Reply:
Why is this hash even needed? It does nothing extra compared to what `MixinSimpleRegistry` does.

Reply:
The only use for this whole hash thing is trying to detect mods that remove from the registry without marking the registry as modified.

The whole thing could be removed as mods that do odd things should start using the api provided to mark the registry as modded. This is here to try and maintain backwards compat where possible.